src/main.rs

```diff
@@ -67,17 +69,90 @@ struct Args {
     test: bool,
 
     /// Compress faster
-    #[clap(short = '1', long)]
+    #[clap(short = '1', long, conflicts_with = "best")]
     fast: bool,
 
     /// Compress better
-    #[clap(short = '9', long)]
+    #[clap(short = '9', long, conflicts_with = "fast")]
     best: bool,
 
+    /// Set compression level
+    #[clap(skip)]
+    compression_level: usize,
+
+    /// Use suffix SUF on compressed files
+    #[clap(short = 'S', long, default_value = "")]
+    suffix: String,
+
+    /// Make rsync-friendly archive
+    #[clap(long)]
+    rsyncable: bool,
+
+    /// Synchronous output (safer if system crashes, but slower)
+    #[clap(long)]
+    synchronous: bool,
+
     #[clap(value_name = "FILE")]
     files: Vec<PathBuf>,
 }
 
 fn main() {
-    let _args = Args::parse();
+    // Parse and validate arguments.
+    let (level, cleaned_args) = clean_args(env::args());
+    let mut args = Args::parse_from(cleaned_args.iter());
+    args.compression_level = match (args.fast, args.best, level) {
+        (_, true, _) => 9,
+        (_, _, Some(0)) => {
+            println!("{}", Args::into_app().error(ErrorKind::ArgumentNotFound, "invalid option -- '0'"));
+            return;
+        }
+        (_, _, Some(d)) => d,
+        (true, _, None) => 1,
+        _ => DEFAULT_COMPRESSION_LEVEL,
+    };
+}
+
+/// Consume an iterator of arguments into the final level flag `(-1, -2, ..., -9)`, if
+/// present, and its arguments after removal.
+///
+/// For example, `["-1dc3", "--fast"]` becomes `(Some(3), ["-dc", "--fast"])`.
+/// 
+/// This function makes no guarantee as to whether the returned level flag is valid in
+/// the program context. In the case of gzip, -0 is invalid, and must be handled
+/// separately.
+fn clean_args<I: IntoIterator<Item = String>>(args: I) -> (Option<usize>, Vec<String>) {
+    let (mut level, mut cleaned_args) = (None, vec![]);
+    args.into_iter().for_each(|s| {
+        let mut cleaned = String::new();
+        if s.starts_with("-") && !s.starts_with("--") {
+            s.chars().for_each(|c| {
+                if c == '=' {
+                    return;
+                }
+                if let Some(d) = c.to_digit(10) {
+                    level = Some(d as usize);
+                } else {
+                    cleaned.push(c);
+                }
+            });
+            if !["", "-"].contains(&cleaned.as_str()) {
```

Comment:
We shouldn't merely ignore these strings. If we have `gzip ''` then the `''` should be treated as an empty file name. Similarly with `gzip -`, except the `-` gets treated as stdin. Also, what if we have `--best -5`? I believe in general `-5` should override but that doesn't seem to happen right now.

A better option (though a bit awkward) is to have `level2`…`level8` flags in the derived struct. This is also a more direct translation of the gzip C code.